homeassistant/helpers/template.py

```diff
@@ -29,12 +30,48 @@
 _SENTINEL = object()
 DATE_STR_FORMAT = "%Y-%m-%d %H:%M:%S"
 
-_RE_NONE_ENTITIES = re.compile(r"distance\(|closest\(", re.I | re.M)
-_RE_GET_ENTITIES = re.compile(
-    r"(?:(?:states\.|(?:is_state|is_state_attr|state_attr|states)"
-    r"\((?:[\ \'\"]?))([\w]+\.[\w]+)|([\w]+))", re.I | re.M
-)
-_RE_JINJA_DELIMITERS = re.compile(r"\{%|\{\{")
+_ENTITY_COLLECT = None
+
+# pylint: disable=invalid-name
+
+
+class async_generate_filter(FilterBuilder):
+    """
+    Filter function generator for entities touched in template.
+
+    Returns Generate a filter function that returns true for any entity IDs
+    passed to it that are used to calculate results of the template.
+
+    Usage:
+      with async_generate_filter() as entities:
+        template1.async_render(variables)
+        template2.async_render(variables)
+    """
+
+    def __init__(self):
+        """Initialise."""
+        super().__init__()
+        self._filter = None
+
+    def __enter__(self) -> Callable[[str], bool]:
+        """Return `self` upon entering the runtime context."""
+        global _ENTITY_COLLECT
```

Comment:
I don't think that this is a good idea.

Reply:
I did start with a thread local here, but since it's single threaded code anyhow (which we can guarantee from the semantics of `async_` and all the methods being used being async) then it is actually safe. We could store the variable on the template, but most of the methods here don't get passed the template anyhow they're bare global methods and besides it's really a stack variable not a instance variable.

There's not really another way to do it, short of passing an accumulator parameter to every method which is going to be a right PITA and isn't actually required anyhow because this method works. :grin:

Reply:
How about changing the signature of render like this:

```
    def async_render(self, variables: TemplateVarsType = None,
                     filter_builder: FilterBuilder = None,
                     **kwargs) -> str:
```

and then accumulating the filter on the passed builder (if present). That way it stops the context wrapper from leaking out of the package. Make it private too.

Reply:
I don't fully understand. Each template is an instance, why can't we save it on there?

Reply:
Take the template:

```
{{ states[states.input_text.myinput.state].state }}
```

At template definition time there's no way to know the list of all possible states that could impact the result of the template, because it's dynamically defined. You'd have to use `*` which would create a load of extra work.

What this change does is whenever we evaluate the template, we keep a list of the states used. So for this example if we ran the template with the `input_test.myinput.state=="sensor.humidity"` we'd end up with `['input_test.myinput', 'sensor.humidity']` on the list. We know the result of the template can't change unless one of those two states change, so we listen for a change on those states. 

If the user then changes `input_test.myinput.state=="sensor.temp"` then the state change event is picked up, the template is reevaluated, and the list becomes  `['input_test.myinput', 'sensor.humidity']`. 

If the user puts in `input_test.myinput.state=="blah invalid"`, because the state name is invalid we end up with `['input_test.myinput']`, for now I hope obvious reasons.

So the list doesn't really belong with the template itself, because the list is really dependent on the inputs to the `async_render` method and friends. It's going to change depending on the input of that method. It's really part of the *output* from the rendering process.

Doing this dynamic computation of the dependent states actually costs essentially nothing, it's just stuff you'd have to look at anyhow to compute the template result, and for your negligible cost you get a completely robust system that only reevaluates the template when it needs to.

Reply:
I'm actually thinking I don't like changing the signature of `async_render`, it's too heavily used and has `**kwargs` too, there's too much potential for issues. I will create a new method:

```
def async_render_collect_entities(
    self, variables: TemplateVarsType = None,
    **kwargs) -> (str, EntityFilter):
```

Also the FilterBuilder class is a bit daft. There's no actual benefit in finalising the filter to a method, you're literally just avoiding two boolean comparisons. It's just as well to have an `EntityFilter` class that gets output from the `entity_filter` method, and does everything the current `entity_filter` does in its `__call__` method. That way you can introspect the result to review the list of included/excluded domains and entities.